draft-ietf-quic-transport.md

```diff
@@ -3488,6 +3489,54 @@ Even if validation fails, an endpoint MAY revalidate ECN on the same path at any
 later time in the connection.
 
 
+#### Exemplary ECN Validation Algorithm {#ecn-alg}
+
+Each time that an endpoint commences sending on a new network path, it
+determines whether the path supports ECN.  If the path supports ECN, the goal is
+to use ECN.  Endpoints might also periodically reassess a path that was
+determined not support ECN.
+
+This section describes one method for testing new paths.  This algorithm is
+intended to show how a path might be tested for ECN support.  Endpoints can
+implement different methods.
+
+The path has an ECN state that is one of "testing", "unknown", "failed", or
+"capable".  In the "testing" and "capable" states the endpoint sends packets
+with an ECT(0) marking; otherwise, the endpoint sends unmarked packets.
+
+To start testing a path, the ECN state is set to "testing" and existing ECN
+counts are remembered as a baseline.  It should not be necessary to disregard
+the effect of packets sent prior to starting testing, though it is necessary for
+a sender to remember what markings were used for every packet that is
+acknowledged; see {{ecn-ack}}.
+
+The testing period runs for a number of packets or round trip times as
+determined by the endpoint.  During this time, packets sent are marked with
+ECT(0).  The goal is to limit the duration of the testing period, but to ensure
+that enough marked packets are sent that it is likely that ECN counts will
+provide a clear indication of how the path treats marked packets.
+
+<!-- Do we need a more concrete recommendation here?  For instance, I might say
+"Endpoints could test with packets that amount to between 1 to 2 times the
+initial congestion window over a period between 1 to 2 times the estimated RTT."
+-->
+
+After the testing period ends, the ECN state for the path becomes "unknown".
+From the "unknown" state, successful validation of the ECN counts an ACK frame
+(see {{ecn-ack}}) causes the ECN state for the path to become "capable", unless
+no marked packet has been acknowledged.
```

Comment:
This extra condition caught me by surprise.  The problem is that if every packet you mark is lost, the ECN counts remain valid, but you have a fair indication that the path is a black hole.  That's not worth belabouring because black holes are exceedingly rare, but the extra condition seems necessary even if we don't provide a whole lot of explanation.

Reply:
Given that black-holing is rare, I'd rather keep simply all packet ECT-marked until a failure is detect. If black-holing is happening, that's actually the easiest case to detect (as you will have a time-out). Moreover if you plan to send more than one cwnd of packets with ECT marks in testing, you will get blocked anyway as you can't send anymore packets without receiving an ack.

Reply:
You will be blocked, but only until the PTO timer fires. There is a question still of what to do afterwards. While it is rare, I suspect we need this fallback as a failsafe.

Reply:
My thinking was that if you get an PTO in testing to go straight to failed and don't sent any ECT markings anymore.

Reply:
I don't think that works without additional constraints or checks as you could exit the testing state before any PTO might occur.

Reply:
I think the idea would be to only have three states: "Unknown/testing" and "capable" where ETC marking is enabled, "failed" where ECT marking is disabled and you stay in "unknown/testing" until the path has be validate for failed.